---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!--You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/master/examples>. -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

# Flip the switch to evaluate all when ready for final render
Final_eval <- FALSE
```

# Genetic Interaction Network IdentifieR (GINIR)

<!-- badges: start -->
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

![](./GINIR_hex_logo-02.png)

GINIR is an R package that leverages data generated by the [Cancer Dependency Map (DepMap) project](https://depmap.org/portal/) to identify cancer cell lines with mutations of interest and perform in-silico genetic knockout screens. A manuscript describing the process is being prepared.

Current DepMap data used by default is version 20Q1, which was downloaded through the DepMap data portal. The data was distributed and used under the terms and conditions of [CC Attribution 4.0 license](https://creativecommons.org/licenses/by/4.0/). 

## Contributions

This repository is being developed and maintained by [Yuka Takemon](https://github.com/ytakemon) at the [Dr. Marco Marra](https://www.bcgsc.ca/labs/marra-lab)'s laboratory at the [Canada's Michael Smith Genome Sciences Centre](https://www.bcgsc.ca/).

Please submit an [issue](https://github.com/ytakemon/GINIR/issues) for all questions regarding GINIR.

<!-- Below this line are default template stuff that will be updated as the package comes together -->

## Installation

You can install GINIR from [GitHub](https://github.com) with:

``` r
install.packages("devtools")
devtools::install_github("ytakemon/GINIR")
```

## Workflow

1. Select mutant cell lines that carry mutations in the gene of interest and control cell lines. 
    + (optional specifications) disease type, disease subtype, amino acid change
1. Determine differential expression between mutant and control cell line groups.
    + (optional but recommended)
1. Perform *in silico* genetic screen.
1. Visualize results

## Example: Identifying *ARID1A* genetic interactions

*ARID1A* encodes a member of the chromatic remodeling SWItch/Sucrose Non-Fermentable (SWI/SNF) complex and is a frequently mutated gene in cancer. It is known that *ARID1A* and its homolog, *ARID1B*, are synthetic lethal to one another: The dual loss of ARID1A and its homolog, ARID1B, in a cell is lethal; however, the loss of either gene alone is not  ([Helming et al., 2014](https://doi.org/10.1038/nm.3480)). This example will demonstrate how we can identify synthetic lethal interactors of *ARID1A* using `GINIR` and predict this known interaction.

For this example you will need to call the following libraries

```{r libraries, eval = Final_eval}
library(GINIR)
library(tidyverse)
```

### Exploring cell lines 

One way to explore cell lines that are available in DepMap is through their [portal](https://depmap.org/portal/). However, there are some simple built-in methods in GINI to provide users with a way to glimpse the data using the series of `list_available` functions: `list_available_mutations()`, `list_available_cancer_types()`, `list_available_cancer_subtypes()`

Current DepMap data used by default is version 20Q1, which contains whole-genome sequencing or whole-exome sequencing annotations for `1775` cancer cell lines (`1270` cell lines with RNA-seq data, `378` cell lines with quantitative proteomics data, and `739` cell lines with CRISPR-Cas9 knockout screen data)

```{r listing_available_muts, eval = FALSE}
# Find hotspot mutations in TP53
list_available_mutations(Gene = "TP53", Is_hotspot = TRUE) 
```

```{r listing_available_cancer_types, eval = Final_eval}
# List all available cancer types
list_available_cancer_types()

# List all available cancer subtypes
list_available_cancer_subtypes(input_disease = "Lung Cancer")
```

### Selecting mutant and control cell line groups

As default `select_cell_lines()` will identify cancer cell lines with loss-of-function alterations in the gene specified and group them into six different groups. 

Loss-of-function alterations include variants that are annotated as: `"Nonsense_Mutation", "Frame_Shift_Ins", "Splice_Site", "De_novo_Start_OutOfFrame", "Frame_Shift_Del", "Start_Codon_SNP", "Start_Codon_Del",` and `"Start_Codon_Ins"`. Copy number alterations are also taken into consideration and group as `"Deep_del", "Loss", "Neutral",` or `"Amplified"`.

The cell line groups assigned by default are:

* `Control` cell lines do not harbor any single nucleotide variations (SNVs) or insertions and deletions (InDels) with a neutral copy number (CN).
* `HomDel` cell lines harbor one or more homozygous deleterious SNVs or have deep CN loss.
* `T-HetDel` cell lines harbor two or more heterozygous deleterious SNVs/InDels with neutral or CN loss.
* `HetDel` cell lines harbor one heterozygous deleterious SNV/InDel with neutral CN, or no SNV/InDel with CN loss.
* `Amplified` cell lines harbor no SNVs/InDels with increased CN.
* `Others` cell lines harbor deleterious SNVs with increased CN.


```{r select_groups, eval = Final_eval}
ARID1A_groups <- select_cell_lines(Input_gene = "ARID1A")

# Show number of cell lines in each group 
count(ARID1A_groups, Group)
```
#### Optional filter for specific cancer types

```{r select_groups_option, eval = Final_eval}
# Find pancreatic cancer cell lines with ARID1A mutations
ARID1A_pancr_groups <- select_cell_lines(Input_gene = "ARID1A", Input_disease = "Pancreatic Cancer")

# Show number of cell lines in each group 
count(ARID1A_pancr_groups, Group)
```

### Check for differential expression 

Of the three mutant cancer cell line groups `ARID1A_HomDel`, `ARID1A_T-HetDel`, and `ARID1A_HetDel`, cancer cell lines with `ARID1A_HomDel` mutations are most likely to result in a loss or reduced expression of *ARID1A*. Therefore, we want to check whether cell lines in `ARID1A_HomDel` mutant group have significantly less *ARID1A* RNA or protein expression compared to control cell lines.

```{r Check_expression_rna, eval = Final_eval}
# Select only HomDel and Control cell lines
ARID1A_HomDel_muts_and_ctrls <- ARID1A_groups %>% filter(Group %in% c("ARID1A_HomDel", "Control"))

# Get RNA expression 
ARID1A_HomDel_muts_and_ctrls_rna <- extract_rna_expr(
  Input_samples = ARID1A_HomDel_muts_and_ctrls$DepMap_ID, 
  Input_genes = "ARID1A")
```

Not all cell lines contain RNA and/or protein expression profiles, and not all proteins were detected by mass spectrometer. (Details on data generation can be found on the [DepMap site](https://depmap.org/portal/).)

```{r Check_expression_protein, eval = FALSE}
# Get protein expression
ARID1A_HomDel_muts_and_ctrls_protein <- extract_protein_expr(
  Input_samples = ARID1A_HomDel_muts_and_ctrls$DepMap_ID,
  Input_genes = "ARID1A")

# Produces an error message since ARID1A protein data is not available
```
Using Welch's t-test, we can check to see whether *ARID1A* RNA expression (in TPM) is significantly reduced in `ARID1A_HomDel` cell lines compared to `Controls`. 

```{r Check_expression_rna_stats, eval = Final_eval}
# Append groups and test differential expression
ARID1A_HomDel_muts_and_ctrls_rna <- left_join(ARID1A_HomDel_muts_and_ctrls_rna,
                                              ARID1A_HomDel_muts_and_ctrls %>% select(DepMap_ID, Group))

# T-test 
t.test(ARID1A_8289 ~ Group, ARID1A_HomDel_muts_and_ctrls_rna)

# plot 
ggplot(ARID1A_HomDel_muts_and_ctrls_rna, aes(x = Group, y = ARID1A_8289)) +
  geom_boxplot()
```

### Perform *in silico* genetic screen

After determining cell lines in the `ARID1A_HomDel` group has statistically significant reduction in RNA expression compared to `Control` cell lines, the next step is to perform a *in silico* genetic screen using `screen_results()`. This uses the dependency probabilities (or **"lethality probabilities"**) generated from DepMap's genome-wide CRISPR-Cas9 knockout screen. 

**Lethality probabilities** range from 0.0 to 1.0 and is quantified for each gene knock out in every cancer cell line screened (There are 18,334 genes targeted in 739 cancer cell lines). A gene knock out with a lethality probability of 0.0 indicates a non-essential for the cell line, and a gene knock out with a 1.0 indicates an essential gene (ie. very lethal). Details can be found in [Meyers, R., et al., 2017](https://doi.org/10.1038/ng.3984)

At its core, `screen_results()` performs multiple Mann-Whitney U tests, comparing lethality probabilities of each targeted gene between mutant and control groups. This generates a data frame with the following columns:

* `GeneName_ID` - Hugo symbol with NCBI gene ID
* `GeneNames` - Hugo symbol
* `_median, _mean, _sd, _iqr` - Control and mutant group's median, mean, standard deviation (sd), and interquartile range (iqr) of dependency probabilities. Dependency probabilities range from zero to one, where one indicates a essential gene (ie. KO of gene was lethal) and zero indicates a non-essential gene (KO of gene was not lethal)
* `Pval` - P-value from Mann Whitney U test between control and mutant groups.
* `Adj_pval` - BH-adjusted P-value.
* `log2FC_by_median` - Log2 normalized median fold change of dependency probabilities (mutant / control).
* `log2FC_by_mean` - Log2 normalized mean fold change of dependency probabilities (mutant / control).
* `CliffDelta` - Cliff's delta non-parametric effect size between mutant and control dependency probabilities. Ranges between -1 to 1.
* `dip_pval` - Hartigan's dip test p-value. Tests whether distribution of mutant dependency probability is unimodel. If dip test is rejected (p-value < 0.05), this indicates that there is a multimodel dependency probability distribution and that there may be another factor contributing to this separation.
* `Interaction_score` - Combined value generated from signed p-values: -log10(Pval) \* sign(log2FC_by_median). Negative scores indicate lethal genetic interaction, and positive scores indicate alleviating genetic interaction.

```{r run_screen, eval = FALSE}
ARID1A_mutant_IDs <- ARID1A_groups %>% filter(Group %in% c("ARID1A_HomDel")) %>% pull(DepMap_ID)
ARID1A_control_IDs <- ARID1A_groups %>% filter(Group %in% c("Control")) %>% pull(DepMap_ID)

# This can take several hours depending on number of lines/cores used. Best to run this overnight.
screen_results <- GINI_screen(
  control_IDs = ARID1A_mutant_IDs, 
  mutant_IDs = ARID1A_control_IDs, 
  core_num = 5, # depends on how many cores you have  
  output_dir = "path/to/results/folder/", # Will save your results here as well as in the variable
  test = FALSE) # use TRUE to run a short test to make sure all will run overnight.
```

```{r load_prepared_results, eval = Final_eval, include = FALSE}
# Load prepared ARID1A screen result
data("sample_ARID1A_KO_screen")
screen_results <- sample_ARID1A_KO_screen
```

We can quickly determine whether any lethal genetic interactions were predicted by `GINIR`. We use a `Pval` cut off of 0.05 and rank based on the `Interaction_score`. 

```{r checkout_res, eval = Final_eval}
screen_results %>% 
  filter(Pval < 0.05) %>%
  arrange(-Interaction_score) %>% 
  select(GeneNames:Mutant_median, Pval, Interaction_score) %>% head
```

We immediately see that *ARID1B*, a known synthetic lethal interaction of *ARID1A*, was a the top of this list.

### Visualize screen results

Finally once the *in silico* screen is complete, results can be quickly visualized using `plot_screen()`. Positive genetic interaction scores indicate potential synthetic lethal genetic interactors, and negative scores indicate potential alleviating genetic interactors. 

```{r plot, eval = Final_eval}
# Visualize results, turn on gene labels, 
# and label three genes each that are predicted to have 
# lethal and alleviating genetic interactions, respectively

plot_screen(result_df = screen_results, 
            label_genes = TRUE, 
            label_n = 3)
```


## Session Info 
```{r sessioninfo}
sessionInfo()
```
