---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!--You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/master/examples>. -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# Genetic Interaction Network IdentifieR (GINIR)

<!-- badges: start -->
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

GINIR is an R package that leverages data generated by the [Cancer Dependency Map (DepMap) project](https://depmap.org/portal/) to identify cancer cell lines with mutations of interest and perform in-silico genetic knockout screens. A manuscript describing the process is being prepared.

Current DepMap data used by default is version 20Q1, which was downloaded through the DepMap data portal. The data was distributed and used under the terms and conditions of [CC Attribution 4.0 license](https://creativecommons.org/licenses/by/4.0/). 

## Contributions

This repository is being developed and maintained by [Yuka Takemon](https://github.com/ytakemon) at the [Dr. Marco Marra](https://www.bcgsc.ca/labs/marra-lab)'s laboratory at the [Canada's Michael Smith Genome Sciences Centre](https://www.bcgsc.ca/).

Please submit an [issue](https://github.com/ytakemon/GINIR/issues) for all questions regarding GINIR.

<!-- Below this line are default template stuff that will be updated as the package comes together -->

## Installation

You can install GINIR from [GitHub](https://github.com) with:

``` r
install.packages("devtools")
devtools::install_github("ytakemon/GINIR")
```

## Workflow

1. Select mutant cell lines that carry mutations in the gene of interest and control cell lines. 
    + (optional) specify cancer type 
1. Determine differential expression between mutant and control cell line groups.
1. Perform in silico genetic screen.
1. Visualize results

## Examples screen (coming soon)

This is a basic example which shows you how to solve a common problem:

```{r example, eval=FALSE}
library(GINIR)
## basic example code
```

## Session Info 
```{r sessioninfo}
sessionInfo()
```
