
<!-- README.md is generated from README.Rmd. Please edit that file -->

<!--You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/master/examples>. -->

# Genetic Interaction Network IdentifieR (GINIR)

<!-- badges: start -->

[![License: GPL
v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

GINIR is an R package that leverages data generated by the [Cancer
Dependency Map (DepMap) project](https://depmap.org/portal/) to identify
cancer cell lines with mutations of interest and perform in-silico
genetic knockout screens. A manuscript describing the process is being
prepared.

Current DepMap data used by default is version 20Q1, which was
downloaded through the DepMap data portal. The data was distributed and
used under the terms and conditions of [CC Attribution 4.0
license](https://creativecommons.org/licenses/by/4.0/).

## Contributions

This repository is being developed and maintained by [Yuka
Takemon](https://github.com/ytakemon) at the [Dr. Marco
Marra](https://www.bcgsc.ca/labs/marra-lab)’s laboratory at the
[Canada’s Michael Smith Genome Sciences
Centre](https://www.bcgsc.ca/).

Please submit an [issue](https://github.com/ytakemon/GINIR/issues) for
all questions regarding GINIR.

<!-- Below this line are default template stuff that will be updated as the package comes together -->

## Installation

You can install GINIR from [GitHub](https://github.com) with:

``` r
install.packages("devtools")
devtools::install_github("ytakemon/GINIR")
```

## Workflow

1.  Select mutant cell lines that carry mutations in the gene of
    interest and control cell lines.
      - (optional) specify cancer type
2.  Determine differential expression between mutant and control cell
    line groups.
3.  Perform in silico genetic screen.
4.  Visualize results

## Examples screen (coming soon)

This is a basic example which shows you how to solve a common problem:

``` r
library(GINIR)
## basic example code
```

## Session Info

``` r
sessionInfo()
#> R version 4.0.2 (2020-06-22)
#> Platform: x86_64-centos7-linux-gnu (64-bit)
#> Running under: CentOS Linux 7 (Core)
#> 
#> Matrix products: default
#> BLAS:   /gsc/software/linux-x86_64-centos7/R-4.0.2/lib64/R/lib/libRblas.so
#> LAPACK: /gsc/software/linux-x86_64-centos7/R-4.0.2/lib64/R/lib/libRlapack.so
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#>  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#>  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#>  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#> [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> loaded via a namespace (and not attached):
#>  [1] compiler_4.0.2    magrittr_2.0.2    tools_4.0.2       htmltools_0.5.1.1
#>  [5] yaml_2.2.1        stringi_1.7.5     rmarkdown_2.9     knitr_1.37       
#>  [9] stringr_1.4.0     xfun_0.29         digest_0.6.29     rlang_0.4.12     
#> [13] evaluate_0.14
```
