---
params:
  Target_gene: NULL
title: "20Q1 Automated Report: `r params$Target_gene` Cancer Cell Line Identification"
author: "Yuka Takemon"
output:
  html_document:
    theme: cosmo
    df_print: paged
    css: style.css
    toc: true
    toc_depth: 4
    toc_float: true
editor_options:
  chunk_output_type: console
---

```{r global_options, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, error = FALSE, opts.lable = "kill_prefix", fig.height= 8, fig.width= 15, fig.align="center", progress = FALSE, verbose = TRUE) # Set global rmd setting
```

```{r setup, include= FALSE}
# Define terms --------------------------------------------------------
Local_dir <- "/Where/ever/you/decide/"
Target_gene <- params$Target_gene

##Load library ----------------------------------------------------------------------
if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse, knitr, readxl, DT, ggsignif, ggrepel, gridExtra, kableExtra, broom, ggbeeswarm, ggsci)

## Load data ------------------------------------------------------------------
load(paste0(Local_dir,"Data/20Q1_full.RData"))
source("Scripts/Helper_functions_compiled.R")

## Create Local_directory if not in existence-----------------------------
if(Local_dir.exists(paste0(Local_dir,"Analysis/Pan_Cancer/",Target_gene)) == FALSE){
  Local_dir.create(paste0(Local_dir,"Analysis/Pan_Cancer/",Target_gene))
}

## Check if valid gene name------------------------------
# Target_gene <- "Cow"
if(any(dep_annot$GeneNames == Target_gene) == FALSE){
  stop("Not a valid gene name or gene doesn't exist in data set")
}
```

*Last updated: `r Sys.Date()`*

# Selecting `r Target_gene` mutant cancer cell lines
There are two main datasets to consider to identify the right cell lines for our interest. Mutation calls (contains position, variant type, allele ratios, alteration effect predictions) and copy number.

## `r Target_gene` copy number status

DepMap provides gene level copy number data that is log2 transformed plus a pseudo count of 1. This is generated by mapping targets onto the segment level calls. Segment mean values are equal to log2(relative to copy number + 1).

A diploid cell lines can have the following:

* 2 copies: log2( 2/2 + 1) =  1
* 1 copy: log2( 1/2 + 1) = 0.59
* complete loss (deep deletion), log2( 0/2 + 1) = 0

I have set a threshold to bin cell lines into the following states:

* Neutral: > 0.75 & < 1.25
* Loss: > 0.25 & < 0.75
* Deep deletion: < 0.25
* Amplified: > 1.25
find
```{r}
target_ID <- dep_annot %>% filter(GeneNames == Target_gene) %>% pull(GeneNameID)

target_copy_num <- copy_num %>%
  select(DepMap_ID, all_of(target_ID)) %>%
  filter(DepMap_ID %in% dep$DepMap_ID) %>%
  arrange(DepMap_ID) %>%
  mutate( Status = case_when(
    !!as.name(target_ID) <= 0.25 ~ "Deep_del",
    !!as.name(target_ID) > 0.25 & !!as.name(target_ID) < 0.75 ~ "Loss",
    !!as.name(target_ID) >= 0.75 & !!as.name(target_ID) < 1.25 ~ "Neutral",
    !!as.name(target_ID) >= 1.25 ~ "Amplified"))

target_copy_num %>% group_by(Status) %>% tally %>%
  kable() %>% kable_styling(bootstrap_options = c("striped", "collapsed"))
```

The copy number and copy number status assignment for each cell lines are shown below:

```{r}
target_copy_num %>%
  mutate(!!target_ID := signif(!!as.name(target_ID), 3)) %>%
  datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))
```

***

<br>

## `r Target_gene` mutations

### All `r Target_gene` mutations

For reference here are all observed `r Target_gene` SNV and InDels:

```{r}
target_mut <- mut_calls %>%
  filter((DepMap_ID %in% dep$DepMap_ID) & (Hugo_Symbol %in% Target_gene)) %>%
  mutate(AC_combined = coalesce(CGA_WES_AC, SangerRecalibWES_AC, SangerWES_AC, RNAseq_AC,
                                HC_AC, RD_AC, WGS_AC), #(Alt:REF)
         AC_ref_NULL = grepl(":0", AC_combined)) %>%
  mutate(AC_Variant = case_when(
           .$AC_ref_NULL == "TRUE" ~ "Hom_Mut",
           TRUE ~ "Het_Mut"),
         AC_Variant = paste0(Variant_Classification," ",AC_Variant)) %>% # are there any with 0 contribution from reference?
  select(Hugo_Symbol, Chromosome:Annotation_Transcript, cDNA_Change:COSMIChsCnt, Variant_annotation:AC_Variant) %>%
  arrange(Start_position) %>%
  select(DepMap_ID, everything())

target_mut %>%
  datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))

all_mutations_count_by_sample <- target_mut %>% count(DepMap_ID)
```

### Only damaging `r Target_gene` mutations

```{r}
mut_dels <- target_mut %>% filter(Variant_annotation == "damaging")
num_del_samples <- mut_dels %>% select(DepMap_ID) %>% distinct()
del_mutations_count_by_sample <- mut_dels %>% count(DepMap_ID)
hom_del_muts_by_sample <- mut_dels %>% select(DepMap_ID, AC_ref_NULL) %>%
  distinct() %>%
  filter(AC_ref_NULL == TRUE)
```

There are a total of `r nrow(mut_dels)` damaging mutations in `r nrow(num_del_samples)` samples. The following variants were defined as damaging:

* De_novo_Start_OutOfFrame
* Frame_Shift_Del
* Frame_Shift_Ins
* Nonsense_Mutation
* Splice_Site
* Start_Codon_Del
* Start_Codon_Ins
* Start_Codon_SNP

The `AC_ref_NULL` column asks whether the reference allele was NULL detected by either WES or RNA-seq. A `AC_ref_NULL == TRUE` indicates that no reference allele was detected, therefore likely a homozygous mutant allele.

```{r}
mut_dels %>%
  datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))

multi_mut_dels <- mut_dels %>%
  filter(AC_ref_NULL == FALSE) %>%
  add_count(DepMap_ID) %>% filter(n > 1)
count_sample_multi_mut_dels <- multi_mut_dels %>% select(DepMap_ID) %>% distinct()
```

### Check for Trans-heterozygous mutations

Here we look for cell lines that carry trans-heterozygous `r Target_gene` alterations. We define trans-heterozygous carries to be cell lines carrying **multiple heterozygous mutations**. Looks like there are `r nrow(count_sample_multi_mut_dels)` samples with  Here is the list of samples and their multiple mutations:

```{r}
multi_mut_dels %>%
  arrange(DepMap_ID) %>%
  datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))
```

***

## Cell line group assignment

For each `r nrow(dep)` cell lines, all of which have gone through Achilles screening, I have created the following groups based on # of mutations and copy number status. This is just a suggestion and will work with you to pick the right groups

Here are the criteria for assignment:

* **`r Target_gene` mutant group 1**: (contains hom. damaging mutations in neutral CN) OR (deep CN loss)
* **`r Target_gene` mutant group 2**: (> 1 het. damaging mutations & neutral CN) OR (1 damaging mutation & CN loss)
* **`r Target_gene` mutant group 3**: (1 het. damaging mutation & neutral CN) OR (0 damaging mutations & CN loss)
* **Control group**: (0 mutations & neutral CN)
* **Amplified group**: (0 mutations & CN amplification)
* **<span style="color: red;">Others</span>**: (1 het. damaging mutation & CN amplification)

```{r, }
summary <- sample_annot %>% filter(DepMap_ID %in% dep$DepMap_ID) %>%
  select(DepMap_ID, stripped_cell_line_name, disease, lineage_subtype, primary_or_metastasis) %>%
  left_join(. , all_mutations_count_by_sample, by = "DepMap_ID") %>%
  rename(Total_mutations = n) %>%
  mutate(Total_mutations = case_when(
    is.na(Total_mutations) ~ as.double(0),
    TRUE ~  as.double(Total_mutations))) %>%
  left_join(., del_mutations_count_by_sample, by = "DepMap_ID") %>%
  rename(Del_mutations = n) %>%
  mutate(Del_mutations = case_when(
    is.na(Del_mutations) ~ as.double(0),
    TRUE ~ as.double(Del_mutations)),
    Del_hom_mut = case_when(
      DepMap_ID %in% hom_del_muts_by_sample$DepMap_ID ~ TRUE,
      TRUE ~ FALSE)) %>%
  left_join(., target_copy_num %>% select(DepMap_ID, Status), by = "DepMap_ID") %>%
  rename(CN_status = Status) %>% arrange(-Del_mutations, -Total_mutations, CN_status)

Groups <- summary %>% mutate(
  Group = case_when(
    ((CN_status == "Deep_del") | (Del_hom_mut == TRUE)) ~ paste0(Target_gene,"_mut_1"),
    ((Del_mutations > 1) & (CN_status == "Neutral")) |
      ((Del_mutations == 1) & (CN_status == "Loss")) ~ paste0(Target_gene,"_mut_2"),
    ((Del_mutations == 1) & (CN_status == "Neutral")) |
      ((Del_mutations == 0) & (CN_status == "Loss")) ~ paste0(Target_gene,"_mut_3"),
    ((Total_mutations == 0) & (CN_status == "Neutral")) ~ "Control",
    ((Total_mutations == 0) & (CN_status == "Amplified")) ~ "Amplified",
    ((Del_mutations == 1) & (CN_status == "Amplified")) ~ "Others",
    TRUE ~ "Others")) %>%
  mutate(Group = fct_relevel(Group, "Control", "Amplified", paste0(Target_gene,"_mut_1"),  paste0(Target_gene,"_mut_2"), paste0(Target_gene,"_mut_3"), "Others"))

Groups %>% group_by(Group) %>% tally
Groups %>% datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))
```


## `r Target_gene` RNA & protein expression

Using the assigned groups from the previous section, we can observed whether RNA or protein expression differ between the groups.

```{r}
# Extract protein data if any
extract_target_protein <- NULL
extract_target_protein_supp <- NULL
if(any(protein_nodup$Gene_Symbol %in% Target_gene)){
  extract_target_protein <- protein_nodup %>%
    filter(Gene_Symbol %in% Target_gene) %>%
    select(-c(Protein_Id,Description, Group_ID), ends_with("_TenPx**")) %>%
    pivot_longer(., -c(Gene_Symbol,Uniprot, Uniprot_Acc)) %>%
    mutate(DepMap_ID = map_chr(name, GetDepMapID))

  if(length(names(table(extract_target_protein$Uniprot_Acc))) == 1) {
    target_uniprot <- names(table(extract_target_protein$Uniprot_Acc))
  } else {
    print("contains > 1 uniprot acc. id")
    target_uniprot <- NULL
    extract_target_protein <- NULL
    extract_target_protein_supp <- NULL
  }

  if(!is.null(target_uniprot)){
    extract_target_protein_supp <- extract_target_protein %>%
      filter(str_detect(DepMap_ID, "^TenPx"))
    extract_target_protein <- extract_target_protein %>%
      filter(str_detect(DepMap_ID, "^ACH-") & DepMap_ID %in% dep$DepMap_ID)
    extract_target_protein <- extract_target_protein %>% filter(Uniprot_Acc == target_uniprot)
    extract_target_protein_supp <- filter(extract_target_protein_supp, Uniprot_Acc == target_uniprot)
  }
}

# Extract mRNA data if any
extract_target_rna <- CCLE_exp %>%
  select(DepMap_ID, !!as.name(target_ID)) %>%
  filter(DepMap_ID %in% dep$DepMap_ID)

# Add to summary
Groups <- Groups %>% left_join(., extract_target_rna, by = "DepMap_ID") %>%
  rename(RNA = !!as.name(target_ID)) %>%
  mutate(Group = fct_relevel(Group, "Control", "Amplified", paste0(Target_gene,"_mut_1"),  paste0(Target_gene,"_mut_2"), paste0(Target_gene,"_mut_3"), "Others"))

if(!is.null(extract_target_protein)){
  Groups <- Groups %>%
    left_join(., extract_target_protein %>% select(DepMap_ID, value)) %>%
  rename(protein = value)
}

# Save Groups info:
write_csv(Groups,
          path = paste0(Local_dir,"Analysis/Pan_Cancer/",
                          Target_gene,"/Groups.csv"))
```

First we take a look at how many samples have RNA & protein data available:

```{r}
if(!is.null(extract_target_protein)){
  show <- Groups %>% group_by(Group) %>%
    summarise(
      Total = n(),
      RNA = sum(!is.na(RNA)),
      Protein = sum(!is.na(protein)),
      .groups = "drop") %>%
    datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))

  pval_anov_rna <- aov(RNA ~ Group, Groups) %>% tidy() %>% filter(term == "Group") %>% pull(p.value) %>% signif(., 3)
  pval_anov_prot <- aov(protein ~ Group, Groups) %>% tidy() %>% filter(term == "Group") %>% pull(p.value) %>% signif(., 3)
} else {
  show <- Groups %>% group_by(Group) %>%
      summarise(
        Total = n(),
        RNA = sum(!is.na(RNA)),
        .groups = "drop") %>%
      datatable(extensions = "Buttons",
            rownames = FALSE,
            options = list(dom = "Blfrtip",
                           lengthMenu = list(c(10, 50, -1), c(10,50,"All")),
                           buttons = c("copy", "csv")))

  pval_anov_rna <- aov(RNA ~ Group, Groups) %>% tidy() %>% filter(term == "Group") %>% pull(p.value) %>% signif(., 3)
  pval_anov_prot <- NULL
}

show
```

ANOVA comparing RNA `r if(!is.null(pval_anov_prot)){"and protein"}` expression shows a p-value of `r pval_anov_rna` `r if(!is.null(pval_anov_prot)){" and "}` `r if(!is.null(pval_anov_prot)){pval_anov_prot}` `r if(!is.null(pval_anov_prot)){", respectively"}`

```{r Fig1_expr_boxplots, fig.align="center"}
if(!is.null(extract_target_protein)){

  max <- max(Groups$protein, na.rm= T)
  min <- min(Groups$protein, na.rm= T)

  protein_data <- Groups %>%
    filter(Group != "Others",
           !is.na(protein)) %>%
    mutate(Group = fct_relevel(Group, "Control", "Amplified", paste0(Target_gene,"_mut_1"),  paste0(Target_gene,"_mut_2"), paste0(Target_gene,"_mut_3"), "Others"))

  protein_plot <- ggplot(protein_data, aes(x = Group, y = protein)) +
    geom_boxplot()+
    labs(title = paste(Target_gene, "protein"),
       y = "normalized expression",
       x = "") +
    theme_bw()+
    ylim(min, max + 4) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size=20))+
    geom_signif_customHSD(protein_data, "protein")
}

max <- max(Groups$RNA, na.rm= T)
min <- min(Groups$RNA, na.rm= T)

rna_plot <- Groups %>% filter(Group != "Others") %>%
  mutate(Group = fct_relevel(Group, "Control",
                             "Amplified",
                             paste0(Target_gene,"_mut_1"),  paste0(Target_gene,"_mut_2"),
                             paste0(Target_gene,"_mut_3"), "Others")) %>%
  ggplot(., aes(x = Group, y = RNA)) +
    geom_boxplot()+
    labs(title = paste(Target_gene, "RNA"),
       y = "TPM",
       x = "") +
    theme_bw()+
    ylim(0, max+5)+
    theme( axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size=20))+
   geom_signif_customHSD(Groups %>% filter(Group != "Others"), "RNA")

if(!is.null(extract_target_protein)){
  grid.arrange(rna_plot, protein_plot, ncol = 2)
} else {
  print(rna_plot)
}

# Save plot
if(file.exists(paste0(Local_dir,"Analysis/Pan_Cancer/",Target_gene,"/Expression_byGroup.pdf")) == FALSE){

  if(!is.null(extract_target_protein)){
    pdf(paste0(Local_dir,"Analysis/Pan_Cancer/",Target_gene,"/Expression_byGroup.pdf"), height = 6, width = 10)
    grid.arrange(rna_plot, protein_plot, ncol = 2)
    dev.off()
  } else {
    pdf(paste0(Local_dir,"Analysis/Pan_Cancer/",Target_gene,"/Expression_byGroup.pdf"), height = 6, width = 10)
    print(rna_plot)
    dev.off()
  }
}
```
**p-values: `* < 0.05`, `** < 0.01`, `*** < 0.001`**
